### PR TITLE
chore: Basic object tracking

### PR DESCRIPTION
**Pull request by sfc-gh-jcieslak**
## Changes
- All proposals for basic object tracking tested
- Added functions (and tested them) that allow us to use Golang's context to chosen usage tracking

## Next pr
- https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3205#discussion_r1850109561

### CODE REVIEW

File: pkg/sdk/client_integration_test.go [hunk 1]
```diff
@@ -132,3 +135,72 @@ func TestClient_NewClientDriverLoggingLevel(t *testing.T) {
...
+
+	assertQueryMetadata := func(t *testing.T, queryId string) {
+		t.Helper()
+		result, err := client.QueryUnsafe(context.Background(), fmt.Sprintf("SELECT QUERY_ID, QUERY_TEXT FROM TABLE(INFORMATION_SCHEMA.QUERY_HISTORY(RESULT_LIMIT => 2)) WHERE QUERY_ID = '%s'", queryId))
+		require.NoError(t, err)
+		require.Len(t, result, 1)
+		require.Equal(t, queryId, *result[0]["QUERY_ID"])
+		var parsedMetadata map[string]string
+		queryText := (*result[0]["QUERY_TEXT"]).(string)
+		queryMetadata := strings.Split(queryText, fmt.Sprintf("--%s", DashboardTrackingPrefix))[1]
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/client_integration_test.go:162]: nit: both `toTrackingComment` and `fromTrackingComment` could be extracted as prod code functions (I mean, if we have prod function to create a comment we could also have the method to do it the other way round)

File: pkg/sdk/testint/basic_object_tracking_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	sessionId, err := client.ContextFunctions.CurrentSession(ctx)
+	require.NoError(t, err)
+
+	queryIdChan := make(chan string, 1)
+	metadata := `{"comment": "some comment"}`
+	_, err = client.QueryUnsafe(gosnowflake.WithQueryIDChan(ctx, queryIdChan), fmt.Sprintf(`SELECT 1; --%s`, metadata))
+	require.NoError(t, err)
+	queryId := <-queryIdChan
+
+	result, err := client.QueryUnsafe(ctx, fmt.Sprintf("SELECT QUERY_ID, QUERY_TEXT FROM TABLE(INFORMATION_SCHEMA.QUERY_HISTORY_BY_SESSION(SESSION_ID => %s, RESULT_LIMIT => 2)) WHERE QUERY_ID = '%s'", sessionId, queryId))
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/basic_object_tracking_integration_test.go:73]: nit: extract this query into helper func for easier reuse (you already use it at least five times)

File: pkg/sdk/testint/basic_object_tracking_integration_test.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+// https://docs.snowflake.com/en/sql-reference/parameters#query-tag
+func TestInt_ContextQueryTags(t *testing.T) {
+	client := testClient(t)
+	ctx := context.Background()
+
+	sessionId, err := client.ContextFunctions.CurrentSession(ctx)
+	require.NoError(t, err)
+
+	queryTag := "some query tag"
+	require.NoError(t, client.Parameters.SetSessionParameterOnAccount(ctx, sdk.SessionParameterQueryTag, queryTag))
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/basic_object_tracking_integration_test.go:24]: this does not invoke https://docs.snowflake.com/en/sql-reference/sql/alter-session but alter account and in cleanup we clean a session. Let's use ALTER SESSION SET instead of SetSessionParameterOnAccount.

File: pkg/sdk/testint/basic_object_tracking_integration_test.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/snowflakedb/gosnowflake"
+	"github.com/stretchr/testify/require"
+	"strings"
+	"testing"
+)
+
+// Research for basic object tracking done as part of SNOW-1737787
+
+// https://docs.snowflake.com/en/sql-reference/parameters#query-tag
+func TestInt_ContextQueryTags(t *testing.T) {
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/basic_object_tracking_integration_test.go:16]: what about the precedence test (i.e. set on ACC -> user -> session -> gosnowflake.WithQueryTag)? we are not going with this method now, so please leave a TODO with issue number at least

File: pkg/sdk/client_integration_test.go [hunk 5]
```diff
@@ -3,7 +3,10 @@ package sdk
 import (
 	"context"
 	"database/sql"
+	"encoding/json"
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/client_integration_test.go:6]: let's add an example usage in one of the resources and add one test from the terraform perspective (please keep in mind some nice helper func for tracking metadata creation)

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: pkg/sdk/testint/basic_object_tracking_integration_test.go (hunk 3 above)
Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/basic_object_tracking_integration_test.go:24]: that's the reason why we currently have these fails on the CI:
```
users_acceptance_test.go:54: Step 1/2 error: Check failed: check 3/35 error:
        data.snowflake_users.test parameters assertion [33/59]: failed with error: Attribute 'users.0.parameters.0.query_tag.0.value' expected "", got "some query tag"
 ```

File: pkg/sdk/testint/basic_object_tracking_integration_test.go (hunk 4 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/testint/basic_object_tracking_integration_test.go:16]: A query tag is a regular parameter that goes under the regular hierarchy of precedence; the only nuance with this one is that we can set it on the query level, which was not documented. I tested it here by setting parameter on session level (the "lowest possible" one) to see if setting it on query level overrides it (it does). I'll describe it at the top, so it's clear.

Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/basic_object_tracking_integration_test.go:16]: it would be nice to have at least one more level tested because then it's a clear answer when someone asks why not use query tags

File: pkg/sdk/client.go [hunk 6]
```diff
@@ -264,20 +265,27 @@ func (c *Client) Close() error {
...
 }
 
-type snowflakeAccountLocatorContext string
+type ContextKey string
 
 const (
-	snowflakeAccountLocatorContextKey snowflakeAccountLocatorContext = "snowflake_account_locator"
+	SnowflakeAccountLocatorContextKey ContextKey = "snowflake_account_locator"
+	MetadataContextKey                ContextKey = "metadata"
+	DashboardTrackingPrefix                      = "dashboard_tracking_"
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/client.go:273]: Done

File: pkg/sdk/client_integration_test.go [hunk 7]
```diff
@@ -132,3 +135,72 @@ func TestClient_NewClientDriverLoggingLevel(t *testing.T) {
 		assert.Equal(t, "trace", gosnowflake.GetLogger().GetLogLevel())
 	})
 }
+
+func TestClient_AdditionalMetadata(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/client_integration_test.go:139]: For now, moved only newly added ones

File: pkg/sdk/client_integration_test.go [hunk 8]
```diff
@@ -132,3 +135,72 @@ func TestClient_NewClientDriverLoggingLevel(t *testing.T) {
...
+	client := defaultTestClient(t)
+
+	// needed for using information_schema
+	databaseId := randomAccountObjectIdentifier()
+	require.NoError(t, client.Databases.Create(context.Background(), databaseId, &CreateDatabaseOptions{}))
+	t.Cleanup(func() {
+		require.NoError(t, client.Databases.Drop(context.Background(), databaseId, &DropDatabaseOptions{}))
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/client_integration_test.go:146]: Removed, in `testint` we already have pre-created database, so this is not needed there

File: pkg/sdk/client_integration_test.go (hunk 1 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/client_integration_test.go:162]: moved to `tracking` package

File: pkg/sdk/testint/basic_object_tracking_integration_test.go (hunk 2 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/testint/basic_object_tracking_integration_test.go:73]: moved to newly created `InformationSchemaClient`

File: pkg/sdk/testint/basic_object_tracking_integration_test.go (hunk 3 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/testint/basic_object_tracking_integration_test.go:24]: Changed to session param change instead of account one.

File: pkg/sdk/client_integration_test.go (hunk 5 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/client_integration_test.go]: Done

File: pkg/internal/tracking/query.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+	}
+}
+
+func ParseMetadataFromSql(sql string) (Metadata, error) {
+	parts := strings.Split(sql, fmt.Sprintf("--%s", MetadataPrefix))
+	if len(parts) != 2 {
+		return Metadata{}, fmt.Errorf("failed to parse metadata from sql, incorrect number of parts, expected: 2, got: %d", len(parts))
+	}
+	var metadata Metadata
+	if err := json.Unmarshal([]byte(strings.TrimSpace(parts[1])), &metadata); err != nil {
```
Comment (sfc-gh-jmichalak) [inline pkg/internal/tracking/query.go:24]: Nit: I think that Unmarshal doesn't require the data to be trimmed.

File: pkg/internal/tracking/query_test.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	expectedSql := fmt.Sprintf("%s --%s %s", sql, MetadataPrefix, string(bytes))
+
+	newSql, err := AppendMetadataToSql(sql, metadata)
+	require.NoError(t, err)
+	require.Equal(t, expectedSql, newSql)
+}
+
+func TestParseMetadataFromSql(t *testing.T) {
+	metadata := NewMetadata("123", resources.Account, CreateOperation)
+	sql, err := AppendMetadataToSql("SELECT 1", metadata)
```
Comment (sfc-gh-jmichalak) [inline pkg/internal/tracking/query_test.go:27]: Nit: this unit test depends on `AppendMetadataToSql`, but we can simply use an explicit query here.

File: pkg/resources/schema.go [hunk 11]
```diff
@@ -128,6 +131,7 @@ func Schema() *schema.Resource {
 
 func ImportSchema(ctx context.Context, d *schema.ResourceData, meta any) ([]*schema.ResourceData, error) {
 	log.Printf("[DEBUG] Starting schema import")
+	ctx = tracking.NewContext(ctx, tracking.NewVersionedMetadata(resources.Schema, tracking.ImportOperation))
```
Comment (sfc-gh-jmichalak) [inline pkg/resources/schema.go:134]: Why is this needed here and not in other functions?

File: pkg/sdk/client.go [hunk 12]
```diff
@@ -300,6 +301,19 @@ func (c *Client) queryOne(ctx context.Context, dest interface{}, sql string) err
...
 }
+
+func appendQueryMetadata(ctx context.Context, sql string) string {
+	if metadata, ok := tracking.FromContext(ctx); ok {
+		newSql, err := tracking.AppendMetadataToSql(sql, metadata)
+		if err != nil {
+			log.Printf("[ERROR] failed to append metadata tracking: %v\n", err)
+			return sql
+		}
+		return newSql
```
Comment (sfc-gh-jmichalak) [inline pkg/sdk/client.go:316]: Can we also log here if there's no metadata in the context?

File: pkg/sdk/testint/basic_object_tracking_integration_test.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+	queryTag := "some query tag"
+	require.NoError(t, client.Sessions.AlterSession(ctx, &sdk.AlterSessionOptions{
+		Set: &sdk.SessionSet{
+			SessionParameters: &sdk.SessionParameters{
+				QueryTag: sdk.String(queryTag),
+			},
+		},
+	}))
+	t.Cleanup(func() {
+		_, err = client.QueryUnsafe(ctx, "ALTER SESSION UNSET QUERY_TAG")
```
Comment (sfc-gh-jmichalak) [inline pkg/sdk/testint/basic_object_tracking_integration_test.go:32]: Why not Sessions.AlterSession with unset?

File: pkg/sdk/testint/basic_object_tracking_integration_test.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+		require.NoError(t, err)
+	})
+
+	queryId := executeQueryAndReturnQueryId(t, context.Background(), client)
+
+	result, err := client.QueryUnsafe(ctx, fmt.Sprintf("SELECT QUERY_ID, QUERY_TAG FROM TABLE(INFORMATION_SCHEMA.QUERY_HISTORY_BY_SESSION(SESSION_ID => %s, RESULT_LIMIT => 2)) WHERE QUERY_ID = '%s'", sessionId, queryId))
```
Comment (sfc-gh-jmichalak) [inline pkg/sdk/testint/basic_object_tracking_integration_test.go:38]: We can use the client helper here.

File: pkg/internal/tracking/query.go [hunk 15]
```diff
@@ -0,0 +1,28 @@
+package tracking
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+func AppendMetadataToSql(sql string, metadata Metadata) (string, error) {
```
Comment (sfc-gh-asawicki) [inline pkg/internal/tracking/query.go:9]: nit: simply AppendMetadata and Parse/ExtractMetadata would be enough

File: pkg/resources/schema.go [hunk 16]
```diff
@@ -89,23 +91,24 @@ var schemaSchema = map[string]*schema.Schema{
 // Schema returns a pointer to the resource representing a schema.
 func Schema() *schema.Resource {
 	return &schema.Resource{
-		CreateContext: CreateContextSchema,
-		ReadContext:   ReadContextSchema(true),
-		UpdateContext: UpdateContextSchema,
-		DeleteContext: DeleteContextSchema,
+		CreateContext: CommonCreateWrapper(resources.Schema, CreateContextSchema),
```
Comment (sfc-gh-asawicki) [inline pkg/resources/schema.go:94]: let's add a check in one of the acceptance tests (or dedicated one) that after adding these, we see the tracking info in the query history. can be added in a separate PR

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: pkg/sdk/testint/basic_object_tracking_integration_test.go (hunk 4 above)
Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/basic_object_tracking_integration_test.go:16]: What about this one?

File: pkg/internal/tracking/query.go (hunk 9 above)
Comment (sfc-gh-jcieslak) [reply pkg/internal/tracking/query.go:24]: I would leave it just to be sure

File: pkg/internal/tracking/query_test.go (hunk 10 above)
Comment (sfc-gh-jcieslak) [reply pkg/internal/tracking/query_test.go:27]: done

File: pkg/internal/tracking/query_test.go [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+	require.NoError(t, err)
+
+	parsedMetadata, err := ParseMetadataFromSql(sql)
+	require.NoError(t, err)
+	require.Equal(t, metadata, parsedMetadata)
+}
+
+func TestParseInvalidMetadataKeysFromSql(t *testing.T) {
+	sql := fmt.Sprintf(`SELECT 1 --%s {"key": "value"}`, MetadataPrefix)
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/internal/tracking/query_test.go:36]: fails now

File: pkg/resources/schema.go (hunk 11 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/schema.go:134]: Seems to be leftover from previous approach, removed

File: pkg/sdk/testint/basic_object_tracking_integration_test.go (hunk 13 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/testint/basic_object_tracking_integration_test.go:32]: done

File: pkg/sdk/testint/basic_object_tracking_integration_test.go (hunk 14 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/testint/basic_object_tracking_integration_test.go:38]: Yeah, forgot about that part

File: pkg/acceptance/helpers/information_schema_client.go [hunk 18]
```diff
@@ -0,0 +1,34 @@
+package helpers
+
+import (
+	"context"
+	"fmt"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk"
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/acceptance/helpers/information_schema_client.go:6]: ran pre-push

File: pkg/internal/tracking/context.go [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+	ReadOperation       Operation = "read"
+	UpdateOperation     Operation = "update"
+	DeleteOperation     Operation = "delete"
+	ImportOperation     Operation = "import"
+	CustomDiffOperation Operation = "custom_diff"
+)
+
+type Metadata struct {
+	Version   string                 `json:"version,omitempty"`
+	Resource  resources.ResourceName `json:"resource,omitempty"`
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/internal/tracking/context.go:30]: changed to string but expect resources.Resource on constructor level as discussed

File: pkg/provider/resources/resources.go [hunk 20]
```diff
@@ -1,81 +1,81 @@
 package resources
 
-type resource string
+type ResourceName string
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/provider/resources/resources.go:3]: changed back

File: pkg/resources/common.go [hunk 21]
```diff
@@ -101,3 +104,45 @@ func ImportName[T sdk.AccountObjectIdentifier | sdk.DatabaseObjectIdentifier | s
 
 	return []*schema.ResourceData{d}, nil
 }
+
+func CommonImportWrapper(resourceName resources.ResourceName, importImplementation schema.StateContextFunc) schema.StateContextFunc {
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/resources/common.go:108]: Changed `Common` to `Tracking`, I was thinking about better name, but just left with generic one.

File: pkg/resources/common.go [hunk 22]
```diff
@@ -2,6 +2,9 @@ package resources
 
 import (
 	"context"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/internal/tracking"
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/resources/common.go:5]: ran pre-push

File: pkg/resources/common.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/resources/common.go:108]: I think I would leave them here, it makes sense to group them with things that are already present here. I was looking for a better place like helpers for resources, but we have this for now. Maybe with common delete function etc. it could be moved to a resource_helpers.go (or something similar) file.

File: pkg/sdk/testint/client_integration_test.go [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/snowflakedb/gosnowflake"
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestInt_Client_AdditionalMetadata(t *testing.T) {
+	client := testClient(t)
+
+	metadata := tracking.Metadata{
+		Version:   "v1.0.0",
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/testint/client_integration_test.go:15]: done

File: pkg/sdk/testint/client_integration_test.go [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+		Resource:  "database",
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/testint/client_integration_test.go:16]: Done

File: pkg/sdk/client.go [hunk 25]
```diff
@@ -3,8 +3,8 @@ package sdk
 import (
 	"context"
 	"database/sql"
-	"encoding/json"
 	"fmt"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/internal/tracking"
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/client.go:7]: ran pre-push

File: pkg/internal/tracking/context.go [hunk 26]
```diff
@@ -0,0 +1,57 @@
+package tracking
+
+import (
+	"context"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/provider/resources"
+)
+
+const (
+	ProviderVersion string = "v0.98.0" // TODO(SNOW-1814934): Currently hardcoded, make it computed
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/internal/tracking/context.go:9]: added

File: pkg/internal/tracking/context.go [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+	MetadataPrefix  string = "terraform_provider_usage_tracking"
+)
+
+type key int
+
+var metadataContextKey key
+
+type Operation string
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/internal/tracking/context.go:17]: I think we need to agree how the enums should look like, in this case I would leave it as is because that is how the majority of our enums look like. Also, like with resources.Resource I would need to hold it inside the struct as plain string.

File: pkg/internal/tracking/context.go [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+	"context"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/provider/resources"
+)
+
+const (
+	ProviderVersion string = "v0.98.0" // TODO(SNOW-1814934): Currently hardcoded, make it computed
+	MetadataPrefix  string = "terraform_provider_usage_tracking"
+)
+
+type key int
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/internal/tracking/context.go:13]: Done

File: pkg/internal/tracking/query.go (hunk 15 above)
Comment (sfc-gh-jcieslak) [reply pkg/internal/tracking/query.go:9]: Renamed

File: pkg/sdk/testint/basic_object_tracking_integration_test.go (hunk 4 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/testint/basic_object_tracking_integration_test.go:16]: added user level

File: pkg/sdk/client.go (hunk 12 above)
Comment (sfc-gh-jcieslak) [reply pkg/sdk/client.go:316]: I'm wondering if that's necessary now (could be added later) because it would log this message a lot

File: pkg/resources/schema.go (hunk 16 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/schema.go:94]: Added in desc for as TODO in next pr

**Review by sfc-gh-jmichalak — APPROVED** [review]

File: pkg/sdk/client.go [hunk 29]
```diff
@@ -277,7 +277,8 @@ func (c *Client) exec(ctx context.Context, sql string) (sql.Result, error) {
 		log.Printf("[DEBUG] sql-conn-exec-dry: %v\n", sql)
 		return nil, nil
 	}
-	ctx = context.WithValue(ctx, snowflakeAccountLocatorContextKey, c.accountLocator)
+	ctx = context.WithValue(ctx, SnowflakeAccountLocatorContextKey, c.accountLocator)
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/client.go:280]: Done
